File: clang/test/Analysis/taint-tester.m
// RUN: %clang_cc1  -analyze -analyzer-checker=experimental.security.taint,debug.TaintTest %s -verify

#import <stdarg.h>

@interface NSString
- (NSString *)stringByAppendingString:(NSString *)aString;
@end
extern void NSLog (NSString *format, ...);
extern void NSLogv(NSString *format, va_list args);

void TestLog (NSString *format, ...);
void TestLog (NSString *format, ...) {
    va_list ap;
    va_start(ap, format);
    NSString *string = @"AAA: ";
    
    NSLogv([string stringByAppendingString:format], ap);
    
    va_end(ap);
}

File: clang/test/Analysis/taint-tester.cpp
// RUN: %clang_cc1  -analyze -analyzer-checker=alpha.security.taint,debug.TaintTest %s -verify
// expected-no-diagnostics

typedef struct _FILE FILE;
typedef __typeof(sizeof(int)) size_t;
extern FILE *stdin;
typedef long ssize_t;
ssize_t getline(char ** __restrict, size_t * __restrict, FILE * __restrict);
int printf(const char * __restrict, ...);
int snprintf(char *, size_t, const char *, ...);
void free(void *ptr);

struct GetLineTestStruct {
  ssize_t getline(char ** __restrict, size_t * __restrict, FILE * __restrict);
};

void getlineTest(void) {
  FILE *fp;
  char *line = 0;
  size_t len = 0;
  ssize_t read;
  struct GetLineTestStruct T;

  while ((read = T.getline(&line, &len, stdin)) != -1) {
    printf("%s", line); // no warning
  }
  free(line);
}

class opaque;
void testOpaqueClass(opaque *obj) {
  char buf[20];
  snprintf(buf, 20, "%p", obj); // don't crash trying to load *obj
}

